docs/architecture.md

```diff
@@ -0,0 +1,223 @@
+
+* **Entity**: A simple unique integer ID linked to the existance of a 'thing'.
+* **Component**: An (optional) property of an entity. Usually a struct stored in a contiguous container that associates it to an Entity.
+* **System**: A function that operates on entities that have a specific set of components. Best described as 'rules' that determine how some components change over time.
+
+Entities are still just integers, but are treated as **compositions** of different types of components. Components represent what an entity **has or is made of**, which determines what operations can be done to it by systems. The components themselves don't describe any behaviour.
+
+As a simple example: any entity with a *Mass component* is affected by the *Gravity system*, any entity with the *Legs component* affected by the *Walk system*. *Mass* or *Legs* don't directly imply any behaviour; it's the world's job to add meaning to them.
+
+Components **never inherit each other**, and all common properties are split into separate components.
+
+As another example, a Planet can be represented by having a Mass, Radius, Atmosphere, and Solid surface component. A Star can share the same Mass, Radius, and Atmosphere, but with a Luminance and Plasma surface component.
+
+### Encapsulation?
+
+Encapsulation is not needed, since it is 100% certain which components a system will read from and write to. Systems are very much analagous to real-world laws of the physics. The real world won't tell a falling apple to update; it's not the apple's responsibility to move itself. The world doesn't care that the apple is even an apple; gravity will affect any object with mass regardless of what it is.
+
+### Updating the world
+
+Most mainstream game engines update the world per-object like this:
+
+```
+void world_update(world) {
+    for (each object in world) {
+        object.update()
+    }
+}
+
+```
+
+An ECS engine will more resemble this:
+
+```
+void world_update(world) {
+    gravity_system(world);
+    move_system(world);
+}
+
+void gravity_system(world) {
+    for (each entity with mass and velocity) {
+        // accelerate the masses together
+    }
+}
+
+void move_system(world) {
+    for (each entity with velocity and position) {
+        position += velocity
+    }
+}
+
+...
+```
+
+Systems are called in a specific order to update the entire world to the next frame. Some systems may depend on each other, and other systems can run in parallel.
+
+The EnTT library provides many features making iteration convenient.
+
+### Responsibility of Systems
+
+Ideally, components are raw data that don't describe any behaviour. However, it's acceptable for them to have constructors, destructors, member functions, and RAII members as long as components are only modified by very specific systems. This includes creating and destroying components. Many systems reading the same component is okay.
```

Comment:
Maybe mention the components aren't inherently protected from cross thread access. The same rules apply for components for multiple writers / readers as in any other framework.